smtp.go

```diff
@@ -2,209 +2,334 @@ package maddy
 
 import (
 	"bytes"
+	"crypto/tls"
 	"errors"
 	"fmt"
 	"io"
+	"log"
+	"net"
 	"os"
-	"strings"
-	"time"
-
-	"github.com/emersion/go-msgauth"
-	pgppubkey "github.com/emersion/go-pgp-pubkey"
-	pgppubkeylocal "github.com/emersion/go-pgp-pubkey/local"
-	pgplocal "github.com/emersion/go-pgpmail/local"
-	smtppgp "github.com/emersion/go-pgpmail/smtp"
+	"strconv"
+
 	"github.com/emersion/go-smtp"
-	smtpproxy "github.com/emersion/go-smtp-proxy"
-	"github.com/emersion/go-smtp/backendutil"
-	"github.com/mholt/caddy/caddyfile"
-	"golang.org/x/crypto/openpgp"
+	"github.com/emersion/maddy/config"
+	"github.com/emersion/maddy/module"
 )
 
-func newSMTPServer(tokens map[string][]caddyfile.Token, hostname string, lmtp bool) (server, error) {
-	be, err := newUpstream(tokens)
-	if err != nil {
-		return nil, err
-	}
+type SMTPUser struct {
+	endp      *SMTPEndpoint
+	anonymous bool
+	username  string
+	state     *smtp.ConnectionState
+}
 
-	if tokens, ok := tokens["auth"]; ok {
-		var err error
-		be, err = newSMTPAuth(caddyfile.NewDispenserTokens("", tokens), be)
-		if err != nil {
-			return nil, err
+func (u SMTPUser) Send(from string, to []string, r io.Reader) error {
+	ctx := module.DeliveryContext{}
+	ctx.From = from
+	ctx.To = to
+	ctx.Anonymous = u.anonymous
+	ctx.AuthUser = u.username
+	ctx.SrcHostname = u.state.Hostname
+	ctx.SrcAddr = u.state.RemoteAddr
+	ctx.OurHostname = u.endp.domain
+	ctx.Ctx = make(map[string]interface{})
+
+	currentMsg := r
+	var buf bytes.Buffer
```

Comment:
Can you add a TODO to run filters in parallel?

Maybe this will require filters to return a `io.WriteCloser` instead, to ensure they get a `ctx` updated by the previous filter.

Reply:
The pipeline is strictly linear. I'm not sure how parallelism would look like here at all since every next step can depend on previous.

Reply:
The idea would be to make the signature be:

```
Process(ctx, out io.WriteCloser) (io.WriteCloser, error)
```

This returns immediately. The incoming message will be written to the returned `io.WriteCloser` by maddy. The filtered message will be written to `out` by the filter.

This allows to chain calls to `Process`. The `openpgp` package uses the same interface.

Reply:
Alright. Will add TODO comment for it.

Reply:
(BTW this is more about memory consumption than latency -- if you have a 10MB e-mail with attachments you need to keep in memory and copy around it'll be pretty intensive)